src/content/blog/astro-4150.mdx

```diff
@@ -0,0 +1,184 @@
+  const input = z.object({ email: z.string().email() }).safeParse(json);
+  if (!input.success) return new Response(JSON.stringify(input.error), { status: 400 });
+  // ...50 more lines
+ }
+
+// After
+defineAction({
+  input: z.object({ email: z.string().email() }),
+  handler: ({ name }) => { /* ... */ },
+})
+```
+
+Actions can be called from client-side components and HTML forms. This gives you to flexibility to build apps using any technology: React, Svelte, HTMX, or just plain Astro components. This example calls a newsletter action and renders the result using an Astro component:
+
+```astro
+---
+// src/pages/newsletter.astro
+import { actions } from 'astro:actions';
+
+const result = Astro.getActionResult(actions.newsletter);
+---
+
+{result && !result.error && <p>Thanks for signing up!</p>}
+
+<form method="POST" action={actions.newsletter}>
+  <input type="email" name="email" />
+  <button>Sign up</button>
+</form>
+```
+
+For more information and usage examples, see our [all-new Actions guide](https://docs.astro.build/en/guides/actions).
+
+## Rewrites in i18n fallbacks
+
+A new option was added to [internationalization (i18n) routing](https://docs.astro.build/en/guides/internationalization/) which allows more control over how fallbacks are handled. You can now specify whether a fallback results in a rewrite (a `200` response from the server) or a redirect, the current default.
+
+The fallback system in i18n allows you to define which content is used for a locale when that content is not available in the user's locale. For example your config might look like:
+
+```js
+import { defineConfig } from "astro/config"
+export default defineConfig({
+  i18n: {
+    defaultLocale: "en",
+    locales: ["es", "en", "fr"],
+    fallback: {
+      fr: "es"
+    },
+    routing: {
+      fallbackType: "rewrite"
+    }
+  }
+})
+```
+
+For more information on this new feature, see [our revamped documentation on handling fallbacks in i18n](https://docs.astro.build/en/guides/internationalization/#fallback).
+
+## LibSQL remotes in Astro DB
+
+You can now specify a [LibSQL](https://turso.tech/libsql) server instance as the remote for `@astrojs/db`. This allows you to self-host your own LibSQL server as an alternative to using Astro DB with Astro Studio. This option works with any supported LibSQL protocol.
+
+To use this feature, set the following environment variables:
+
+- `ASTRO_DB_REMOTE_URL`: the connection URL to your LibSQL server.
+- `ASTRO_DB_APP_TOKEN`: the auth token to your LibSQL server.
+
+Enabling this feature allows you to both connect to the LibSQL server in your production environment, it also allows you to use the Astro DB deployment and push (migration) commands with the LibSQL instance. Thanks to [Fryuni](https://github.com/Fryuni) for this amazing contribution!
+
+For more information on using LibSQL servers in Astro DB, [visit our Astro DB guide](https://docs.astro.build/en/guides/astro-db/#libsql).
```

Comment:
We got a [review from an engineer at Turso in the corresponding docs PR](https://github.com/withastro/docs/pull/8955#discussion_r1734713166) and apparently, the right way we should be writing is `libSQL`! Their homepage has `LibSQL` a few times but I believe they're on it.

The structure of one of the sentences was a bit broken as well, so I reworded it a bit.

```suggestion
## libSQL remotes in Astro DB

You can now specify a [libSQL](https://turso.tech/libsql) server instance as the remote for `@astrojs/db`. This allows you to self-host your own libSQL server as an alternative to using Astro DB with Astro Studio. This option works with any supported libSQL protocol.

To use this feature, set the following environment variables:

- `ASTRO_DB_REMOTE_URL`: the connection URL to your libSQL server.
- `ASTRO_DB_APP_TOKEN`: the auth token to your libSQL server.

Enabling this feature allows you to connect to the libSQL server in your production environment and also allows you to use the Astro DB deployment and push (migration) commands with the libSQL instance. Thanks to [Fryuni](https://github.com/Fryuni) for this amazing contribution!

For more information on using libSQL servers in Astro DB, [visit our Astro DB guide](https://docs.astro.build/en/guides/astro-db/#libsql).
```